File: QA-project/backend/src/db/schema/datasets.ts
import {
  pgTable,
  serial,
  integer,
  varchar,
  timestamp,
} from "drizzle-orm/pg-core";
import { relations } from "drizzle-orm";
import { users } from "./users";
import { defects } from "./defects";
import { analyticsResults } from "./analytics-results";
import { moduleRiskScores } from "./module-risk-scores";
import { aiDocuments } from "./ai-documents";
import { aiQueries } from "./ai-queries";
import { reports } from "./reports";

export const datasets = pgTable("datasets", {
  datasetId: serial("dataset_id").primaryKey(),
  userId: integer("user_id")
    .notNull()
    .references(() => users.userId, { onDelete: "cascade" }),
  fileName: varchar("file_name", { length: 255 }).notNull(),
  fileType: varchar("file_type", { length: 50 }).notNull(),
  uploadType: varchar("upload_type", { length: 50 }).notNull(),
  uploadedAt: timestamp("uploaded_at").defaultNow().notNull(),
});

export const datasetsRelations = relations(datasets, ({ one, many }) => ({
  user: one(users, {
    fields: [datasets.userId],
    references: [users.userId],
  }),
  defects: many(defects),
  analyticsResults: many(analyticsResults),
  moduleRiskScores: many(moduleRiskScores),
  aiDocuments: many(aiDocuments),
  aiQueries: many(aiQueries),
  reports: many(reports),
}));

File: QA-project/backend/app/models/datasets.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, func
from sqlalchemy.orm import relationship
from app.database import Base


class Dataset(Base):
    __tablename__ = "datasets"

    dataset_id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False)
    file_name = Column(String(255), nullable=False)
    file_type = Column(String(50), nullable=False)
    upload_type = Column(String(50), nullable=False)
    uploaded_at = Column(DateTime, server_default=func.now(), nullable=False)

    # Relationships
    user = relationship("User", back_populates="datasets")
    defects = relationship("Defect", back_populates="dataset", cascade="all, delete-orphan")
    analytics_results = relationship("AnalyticsResult", back_populates="dataset", cascade="all, delete-orphan")
    module_risk_scores = relationship("ModuleRiskScore", back_populates="dataset", cascade="all, delete-orphan")
    ai_documents = relationship("AIDocument", back_populates="dataset", cascade="all, delete-orphan")
    ai_queries = relationship("AIQuery", back_populates="dataset", cascade="all, delete-orphan")
    reports = relationship("Report", back_populates="dataset", cascade="all, delete-orphan")
